File: 22-generate-parentheses/22-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans =[]
        generator = [["(",1,0]]
        while generator:
            s , l , r = generator.pop()
            if l-r <0 or l>n or r>n:
                continue
            if l == n and r == n:
                ans.append(s)
                continue
            generator.append([s+"(",l+1,r])
            generator.append([s+")",l,r+1])
        return ans
            
        

File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
public:
    vector<string> ans;
    void solver(string res, int open, int closed, int size,int tempsize){
        if(tempsize == (size<<1) ){
        if(open == size && closed == size)
            ans.push_back(res);    
        return;
        }
        if(open >= closed){
            solver(res+"(", open+1, closed,size,tempsize+1 );
        }
        if(open>closed){
            solver(res+")", open, closed + 1,size,tempsize+1);
        }
    }
    vector<string> generateParenthesis(int n) {
        solver("",0,0,n,0);
        return ans;
    }
};